spring-batch-core/src/main/java/org/springframework/batch/core/job/builder/FlowBuilder.java

```diff
@@ -247,29 +248,32 @@ protected Flow flow() {
 	}
 
 	private void doNext(Object input) {
-		if (this.currentState == null) {
+		if (currentState == null) {
 			doStart(input);
+		} else {
+			State next = createState(input);
+			addTransition("COMPLETED", next);
+			addTransition("*", failedState);
+			currentState = next;
 		}
-		State next = createState(input);
-		addTransition("COMPLETED", next);
-		addTransition("*", failedState);
-		this.currentState = next;
 	}
 
 	private void doStart(Object input) {
-		if (this.currentState != null) {
+		if (currentState == null) {
+			currentState = createState(input);
+		} else {
 			doFrom(input);
 		}
-		this.currentState = createState(input);
 	}
 
 	private void doFrom(Object input) {
 		if (currentState == null) {
 			doStart(input);
+		} else {
+			State state = createState(input);
+			tos.put(currentState.getName(), currentState);
+			currentState = state;
```

Comment:
When `currrent == null`, previous `doFrom` logic seems `createState(input)` twice (inside of `doStart()` and `L273`) so I fix it~!